### DISCUSSION TITLE
Use lazyvim as a library

### DISCUSSION THREAD

**Discussion by jackielii** [LazyVim/LazyVim#2929] [Ideas]
I know this is not one of the design goals, but I've been using Lazyvim as a library for a while and it works great.

It's configured like this:

```lua
-- init.lua
require("config.lazy")


-- lua/lsp/lazyvim.lua
vim.api.nvim_create_autocmd("User", {
  group = vim.api.nvim_create_augroup("LazyVim", { clear = true }),
  pattern = "VeryLazy",
  callback = function()
    LazyVim.format.setup() -- setup autoformat on BufWritePre
    LazyVim.news.setup() -- lazyvim news
    LazyVim.root.setup() -- setup root dir
    -- vim.api.nvim_create_user_command("LazyExtras", function()
    --   LazyVim.extras.show()
    -- end, { desc = "Manage LazyVim extras" })
  end,
})

return {
...
}

-- lua/lsp/builtin.lua
return {
  { import = "lsp.lazyvim" },
  { import = "lsp.lspconfig" },
  { import = "lsp.formatting" },
  { import = "lsp.linting" },

  { import = "lazyvim.plugins.extras.linting.eslint" },
  { import = "lazyvim.plugins.extras.formatting.prettier" },
  { import = "lazyvim.plugins.extras.lang.clangd" },
  { import = "lazyvim.plugins.extras.lang.json" },

  { import = "lsp.extras.lang.python" },
  { import = "lsp.extras.lang.go" },
  { import = "lsp.extras.lang.zig" },
  { import = "lsp.extras.lang.sql" },
  { import = "lsp.extras.lang.htmx_html_templ" },
  { import = "lsp.extras.lang.tex" },
  { import = "lsp.extras.lang.tailwind" },
}


-- lua/config/lazy.lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  -- bootstrap lazy.nvim
  -- stylua: ignore
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable",
    lazypath })
end
vim.opt.rtp:prepend(vim.env.LAZY or lazypath)

require("config.options")
require("config.keymaps")
require("config.autocmds")

require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", lazy = true },
    { import = "plugins" },
    { import = vim.env.USE_COC and "lsp.coc" or "lsp.builtin" },
...
```

As you can see, it's pretty close to how Lazyvim setup is doing, but I prefer to not have the Lazyvim's mappings, vim.opts. And in the meantime, I keep all the lsp related setups.

Recently the new release broke this because the `Lazyvim` global was added in [`lua/lazyvim/config/init.lua:1`](https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/init.lua#L1).

I didn't create this as an issue because I think it's not per design. But it would be really nice to have Lazyvim support this kind of use case.

I can see inside Lazyvim, there are clearly several parts that can be pulled in, e.g. formatting, linting, lsp, extra.lang.*

It has saved me a lot of time looking up docs and configure. I just import the lines, or copy and paste into my own config.

**Comment by dpetka2001** [comment]
You can disable the LazyVim keymaps, options and autocmds as well. By keymaps I mean the general ones that are configured in `/lua/config/keymaps.lua` not the ones that come with plugins. But you can definitely change the keymaps of plugins as well.

**Comment by jackielii** [comment]
Today I spend a bit of time syncing my copy with LazyVim, I realised a lot of things have changed, e.g. tsserver => vtsls, neodev.nvim => lazydev etc. I come back to this issue again...

**It's clearer and clearer what I want is Just LazyVim without any of the keymaps.** 

Keep mine updated with LazyVim's core logic is not hard but it's definitely not 0. And I need to keep it updated in order to get the most out of the latest improvements from LazyVim.

I dream to have this setup:

```
-- init.lua
require("config.lazy")

-- lua/config/lazy.lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  -- bootstrap lazy.nvim
  -- stylua: ignore
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable",
    lazypath })
end
vim.opt.rtp:prepend(vim.env.LAZY or lazypath)

-- my own or import lazy
require("config.options")
require("config.keymaps")
require("config.autocmds")

require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", opts = { nokeymap = true} },
    { import = "lsp" },
    { import = "plugins" },
...
})

-- lua/lsp/my-override.lua
...
-- lua/plugins/my-override.lua
...
```

**Comment by dotfrag** [comment]
I tried a similar approach in the past, this worked great for me, `init.lua`:

```lua
-- lazy.nvim
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.uv.fs_stat(lazypath) then
  local lazyrepo = "https://github.com/folke/lazy.nvim.git"
  vim.fn.system({ "git", "clone", "--filter=blob:none", "--branch=stable", lazyrepo, lazypath })
end
---@diagnostic disable-next-line: undefined-field
vim.opt.rtp:prepend(lazypath)

-- LazyVim utils
local lazyvimpath = vim.fn.stdpath("data") .. "/lazy/LazyVim"
if not vim.uv.fs_stat(lazyvimpath) then
  local lazyrepo = "https://github.com/LazyVim/LazyVim.git"
  vim.fn.system({ "git", "clone", "--filter=blob:none", lazyrepo, lazyvimpath })
end
---@diagnostic disable-next-line: undefined-field
vim.opt.rtp:prepend(lazyvimpath)
_G.LazyVim = require("lazyvim.util")
LazyVim.plugin.lazy_file()
```

Then you can continue as usual, and import only what you want:

```lua
require("config.options")
require("config.keymaps")
require("config.autocmds")

require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim" },
    { import = "lazyvim.plugins.extras.util.mini-hipatterns" },

    { import = "plugins" },
    { import = "plugins.lang" },
  },
  install = { colorscheme = { "tokyonight", "habamax" } },
.....
```

**Reply by jackielii** [reply]
Thanks for sharing! Copying the `LazyVim` global and the `lazy_file` setup from yours! It must have been added recently, because I was using a workaround for the LazyFile event

**Reply by jackielii** [reply]
> I tried a similar approach in the past, this worked great for me

So did you move to the stock LazyVim or your custom solution without LazyVim?

**Reply by dotfrag** [reply]
> So did you move to the stock LazyVim or your custom solution without LazyVim?

I initially started using neovim with LunarVim, then AstroNvim, then Nvchad and finally kickstart when I knew what I wanted.

Then I tried LazyVim and found that it matches what I want like 98%.

I had a good run with a custom solution. I started my config from [kickstart](https://github.com/nvim-lua/kickstart.nvim/), then used LazyVim's plugins and utils when needed. Eventually, I realised I wanted too many stuff from LazyVim. My biggest problem were some keybindings. I specifically do not like repeating keymaps, i.e. `<leader>ff`, `<leader>space`, `<leader>qq`. They are not easy on my hands. The rest I like.

So what I ended up doing is remapping the most commonly used repeating keymaps to my liking, and now I'm super happy. I haven't used my previously preferred editor (VSCode) for a while now. You can see my currently used config [here](https://github.com/dotfrag/nvim).

**Reply by jackielii** [reply]
Cool, just gone through your repo. Looks like you're using the stock LazyVim and override keymaps. Are you overriding the LSP keymaps?

**Reply by dotfrag** [reply]
Stock Lazyvim, yes. Nope, I enjoy them, mostly `gd`, `gr`, and `<leader>c*`. Are you having trouble with those?

**Reply by jackielii** [reply]
I migrated from coc.nvim. I already have most of the lsp actions mapped to my own liking. I think it's possible to remap, there were just too many. Maybe I'll give it another try this Christmas :D

**Reply by dotfrag** [reply]
I'm not familiar with coc.nvim but I'm guessing it has its own set of mappings. Personally I find LazyVim to have the best lsp mappings from all other configs I tried. Good luck :)
